### PR TITLE
Hello

### PR DESCRIPTION
**Pull request by guiwoda**
I'm the original author of this microscopic library.

Two years after I made it, I realized that I had two forks and I never received any pull request from either. Worst yet, one of the forks has hundreds of commits above me, and the other is based on it, so hundred plus!

Welcome to _Open Source_. In here, we try to collaborate to make great software.

@d9magai, meet @captn3m0. You two could get along!

### CODE REVIEW

**Comment by guiwoda** [comment]
By the way, I've deleted my copy, so it's up to you guys now. :wink:

**Comment by captn3m0** [comment]
Thanks for the much needed poke @guiwoda. I should have filed this long back.

Things this PR adds:

- Cache support: cache the instance metadata and re-use if available. Its just storing the json file and re-using it as of now.
- Mock support: This might be slightly untested, but the idea is to let the library be run even in testing environments.

Rest all is just tests, psr and other minor things. Let me know if you'd be interested in merging this back, and I'll clear out the namespace issues.

**Comment by d9magai** [comment]
Hi, @guiwoda, @captn3m0.
Thank you for the Pull Request.
I'm very interested in suggestion of you guys!

I want to try cooperate to make great software.
I will merge this.

**Comment by captn3m0** [comment]
Thanks. I'll fix the namespace in a different branch and create a new PR.

File: src/Ec2MetadataGetter.php [hunk 1]
```diff
@@ -80,6 +79,88 @@ class Ec2MetadataGetter
...
 
     /**
+     * Whether to return dummy data or not
+     * @var boolean
+     */
+    private $dummy = false;
+
+    public function __construct($cache_dir = "/tmp")
+    {
+        $this->cache_dir = $cache_dir;
```
Comment (d9magai) [inline src/Ec2MetadataGetter.php]: `$this->cache_dir` looks like undefined in Ec2MetadataGetter.
So, `$this->cache_dir`  can be rewritten freely.

```
[root@b517f2809e82 ec2-metadata]# cat a.php 
<?php
require_once 'vendor/autoload.php';
$client = new \Razorpay\EC2Metadata\Ec2MetadataGetter();
$client->cache_dir = 'hoge';
echo $client->cache_dir . PHP_EOL;

[root@b517f2809e82 ec2-metadata]# php a.php 
hoge
[root@b517f2809e82 ec2-metadata]# 

```
    
I think we need the declaration of private variables `$cache_dir`.
And, I prefer the name of the `$cacheDir` than `$cache_dir`.Don't you think so?

e.g.
```php
     /**
      * request is cached in this directory.
      *
      * @var string
      */
     private $cacheDir;
```

File: src/Ec2MetadataGetter.php [hunk 2]
```diff
@@ -80,6 +79,88 @@ class Ec2MetadataGetter
...
+
+        // Make sure that it is writeable
+        if(!is_writeable($this->cache_dir))
```
Comment (d9magai) [inline src/Ec2MetadataGetter.php]: I prefer 

```
if (...) 
{
    ;
}
else (...)
{
    ;
}
```

than

```
if (...) {
    ;
} else (...) {
    ;
}
```

And, also I prefer

```
foreach (...)
{
    ;
}
```
than
```
foreach (...) {
    ;
}
```

FYI

**Comment by d9magai** [comment]
>  I'll fix the namespace in a different branch and create a new PR.

Thanks. I understand.

**Comment by d9magai** [comment]
Do you remember me?
I decided the future course about this PR.

It is as follows:
1. I will merge this request.
2. I will edit about my pointed out.

**Comment by guiwoda** [comment]
Awesome!

**Comment by captn3m0** [comment]
@d9magai Can you add me as collaborator to the repo? I'm sorry about forgetting this entirely, but I'll try to fix the issues you'd pointed out.

**Comment by d9magai** [comment]
@captn3m0 

> Can you add me as collaborator to the repo? I'm sorry about forgetting this entirely, but I'll try to fix the issues you'd pointed out.

That's a very good idea!
I will do so.

**Pull request merged**